src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -173,35 +173,30 @@ void BatchNorm<InputDataType, OutputDataType>::Backward(
     const arma::Mat<eT>& gy,
     arma::Mat<eT>& g)
 {
-  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+  const arma::mat invVariance = 1.0 / (variance + eps);
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = input.n_rows / size;
 
   g.set_size(arma::size(input));
-  arma::cube gyTemp(const_cast<arma::Mat<eT>&>(gy).memptr(),
-      input.n_rows / size, size, input.n_cols, false, false);
-  arma::cube gTemp(const_cast<arma::Mat<eT>&>(g).memptr(),
-      input.n_rows / size, size, input.n_cols, false, false);
```

Comment:
This will modify `gy` if `gyTemp` is changed somewhere. Probably if someone wants to access `previousLayer.Delta()`, it will not give the actual result, or if it is used in skip connections it may create a problem. Maybe I am not understanding correctly but can you please clarify this?

Reply:
gY is passed as const, If I try modifying it will, it will throw an error anyway (which is not done). Also this prevents creating an auxiliary array(copy) to prevent space allocation.

Reply:
Can you please try this small program?
```cpp
#include<armadillo>

using namespace arma;

void test(const arma::mat& x,
          arma::mat& y)
{
  size_t splitRatio = 2;
  arma::cube xTemp(const_cast<arma::mat&>(x).memptr(), x.n_rows / splitRatio, x.n_cols, splitRatio, false, false);
  xTemp.zeros();
  
  y.set_size(xTemp.n_rows * xTemp.n_cols, xTemp.n_slices);
  for (size_t i = 0; i < xTemp.n_slices; ++i)
  {
    y.col(i) = arma::vectorise(xTemp.slice(i));
  }
}

int main()
{
  arma::mat x = arma::randu(10, 4);
  arma::mat y;
  x.print("x initially:");
  test(x, y);
  x.print("x finally:");
}

```

Reply:
Right, But I didn't change gYTemp anywhere. 
```

#include<armadillo>

using namespace arma;

void test(const arma::mat& x,
          arma::mat& y)
{
  size_t splitRatio = 2;
  arma::cube xTemp(const_cast<arma::mat&>(x).memptr(), x.n_rows / splitRatio, x.n_cols, splitRatio, false, false);

  y.set_size(xTemp.n_rows * xTemp.n_cols, xTemp.n_slices);
  for (size_t i = 0; i < xTemp.n_slices; ++i)
  {
    y.col(i) = arma::vectorise(xTemp.slice(i));
  }
}

int main()
{
  arma::mat x = arma::randu(10, 4);
  arma::mat y;
  x.print("x initially:");
  test(x, y);
  x.print("x finally:");
}
```
This shouldn't be an issue, so If I make sure not to change it should that be issue ?
Also take a look at Convolution layer, Transposed Conv layers and many more.

Reply:
Ohh...wait! I was reading `gTemp` as `gyTemp`. I'm really sorry 😂. Maybe someone else can suggest better what is good practice here. In my opinion, you can leave if it is "known" that it is not modified.

Reply:
No worries.